src/indexes/text/flat_position_map.h

```diff
@@ -0,0 +1,150 @@
+  for O(log n) skip-forward. Each partition stores [4-byte offset, 4-byte
+  cumulative delta] enabling binary search to target positions.
+
+Delta encoding stores position differences not absolutes.
+
+A FlatPositionMapIterator provides sequential iteration and skip-forward with
+minimal state overhead, maintaining cumulative position for delta decoding.
+
+*/
+
+#include <cstddef>
+#include <cstdint>
+#include <map>
+#include <memory>
+
+namespace valkey_search::indexes::text {
+
+// FlatPositionMap format constants
+constexpr size_t kFlatPositionMapHeaderSize = 4;  // Header is 4 bytes (32 bits)
+constexpr uint8_t kPartitionMapEntrySize =
+    8;  // Each partition entry: 8 bytes (4 offset + 4 delta)
+
+// Encoding bit flags (bit 7 indicates position vs field mask bytes)
+constexpr uint8_t kEncodingBitFieldMask =
+    0x80;                                     // Bit 7 = 1 for field mask bytes
+constexpr uint8_t kEncodingValueMask = 0x7F;  // Lower 7 bits for actual value
+
+// Field mask encoding
+constexpr uint8_t kFieldMaskBitsPerByte =
+    7;  // 7 bits per byte (bit 7 is type flag)
+
+// Encoding scheme thresholds
+constexpr uint32_t kSimpleEncodingMaxPosition =
+    256;  // SIMPLE encoding: positions must be < 256
+constexpr uint32_t kBinarySearchThreshold =
+    128;  // Use BINARY_SEARCH when positions > 128
+
+// Forward declarations to avoid circular dependency
+using Position = uint32_t;
+class FieldMask;
+
+// Encoding schemes
+enum class EncodingScheme : uint8_t {
+  SIMPLE = 0,         // 1 byte per position, no field mask (single field)
+  EXPANDABLE = 1,     // Variable bytes with encoding bit
+  BINARY_SEARCH = 2,  // With partition map for skip operations
+  RESERVED = 3
+};
+
+// FlatPositionMap is a compact byte array representation
+// Layout: [Header: 4 bytes][Optional Partition Map][Position/Field Data]
+class FlatPositionMap {
+ public:
+  // Default constructor: empty map
+  FlatPositionMap() : data_(nullptr) {}
+
+  // Constructor from raw pointer: takes ownership
+  explicit FlatPositionMap(char* d) : data_(d) {}
+
+  // Destructor: frees the allocated memory
+  ~FlatPositionMap();
+
+  // Move constructor: transfers ownership, nullifies source
+  FlatPositionMap(FlatPositionMap&& other) noexcept;
+
+  // Move assignment: frees current, transfers ownership, nullifies source
+  FlatPositionMap& operator=(FlatPositionMap&& other) noexcept;
+
+  // Serialize PositionMap to FlatPositionMap
+  static FlatPositionMap SerializePositionMap(
+      const std::map<Position, std::unique_ptr<FieldMask>>& position_map,
+      size_t num_text_fields);
+
+  // Get position count
+  uint32_t CountPositions() const;
+
+  // Get total term frequency
+  size_t CountTermFrequency() const;
+
+  // Access to raw data pointer
+  const char* data() const { return data_; }
+
+ private:
+  char* data_;
+};
+
+// Iterator for FlatPositionMap
+class FlatPositionMapIterator {
```

Comment:
Should we follow the pattern we've used elsewhere where the iterator is declared inside the FlatPositionMap and the caller can simply call GetPositionIterator() on the FlatPositionMap instance?